server/src/main/java/org/elasticsearch/index/codec/tsdb/es819/DISIAccumulator.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.codec.tsdb.es819;
+
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.util.ArrayUtil;
+import org.apache.lucene.util.BitSetIterator;
+import org.apache.lucene.util.FixedBitSet;
+import org.apache.lucene.util.IOUtils;
+import org.elasticsearch.core.SuppressForbidden;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+/**
+ * Fork of {@link org.apache.lucene.codecs.lucene90.IndexedDISI#writeBitSet(DocIdSetIterator, IndexOutput)} but that allows
+ * building jump list iteratively by one docid at a time instead of relying on docidset iterator.
+ */
+final class DISIAccumulator implements Closeable {
+
+    private static final int BLOCK_SIZE = 65536; // The number of docIDs that a single block represents
+
+    private static final int DENSE_BLOCK_LONGS = BLOCK_SIZE / Long.SIZE; // 1024
+    public static final byte DEFAULT_DENSE_RANK_POWER = 9; // Every 512 docIDs / 8 longs
+
+    static final int MAX_ARRAY_LENGTH = (1 << 12) - 1;
+
+    final Directory dir;
+    final IOContext context;
+    final String skipListTempFileName;
+    final IndexOutput disiTempOutput;
+    final byte denseRankPower;
+    final long origo;
+
+    int totalCardinality = 0;
+    int blockCardinality = 0;
+    final FixedBitSet buffer = new FixedBitSet(1 << 16);
+    int[] jumps = new int[ArrayUtil.oversize(1, Integer.BYTES * 2)];
+    int prevBlock = -1;
+    int jumpBlockIndex = 0;
+
+    DISIAccumulator(Directory dir, IOContext context, IndexOutput data, byte denseRankPower) throws IOException {
+        this.dir = dir;
+        this.context = context;
+        this.disiTempOutput = dir.createTempOutput(data.getName(), "disi", context);
+        this.skipListTempFileName = disiTempOutput.getName();
+        this.denseRankPower = denseRankPower;
+
+        this.origo = disiTempOutput.getFilePointer(); // All jumps are relative to the origo
+        if ((denseRankPower < 7 || denseRankPower > 15) && denseRankPower != -1) {
```

Comment:
nit: Move this check above dir.createTempOutput to avoid leaks if the check fails.